### PR TITLE
Concat tarred ds fixes, additional config parameters

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?

Adds ConcatDataset features to ASR configs and fixes bugs

**Collection**: [Note which collection this PR will affect]
ASR

# Changelog 
- All parameters are now optional
- One needs to pass train_ds.is_concat=True  to activate
- The datasets become a list over the possible current list
- if the concat_sampling_technique = random, one needs to specify concat_sampling_probabilities, one per dataset

# Usage
In the config file
model.train_ds.is_concat=True \
model.train_ds.concat_sampling_technique=random \.  # Default: temperature
# model.train_ds.concat_sampling_temperature=5 \.  # Default: 5
model.train_ds.concat_sampling_probabilities=[0.5,0.5] \.  # Default: none

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x ] New Feature
- [ x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — DISMISSED** [review]

**Comment by titu1994** [comment]
@bmwshop please look into test failure

**Comment by titu1994** [comment]
@bmwshop please look into this

File: nemo/collections/asr/models/ctc_bpe_models.py [hunk 1]
```diff
@@ -177,8 +161,10 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
                 )
         if hasattr(dataset, 'collate_fn'):
             collate_fn = dataset.collate_fn
-        else:
+        elif hasattr(dataset.datasets[0], 'collate_fn'):
             collate_fn = dataset.datasets[0].collate_fn
+        else:
+            collate_fn = dataset.datasets[0].datasets[0].collate_fn
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_bpe_models.py:167]: Lol what's happening here

Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:167]: We need much more docstring to explain this

**Comment by Kipok** [comment]
@bmwshop Can we expedite this PR completion? It fixes an actual bug in the usage of concat dataset with tarred files, so unless we merge this, our users are unable to use concat datasets with tarred files.

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 1 above)
Comment (bmwshop) [reply nemo/collections/asr/models/ctc_bpe_models.py:119]: added comments. This is to support concat datasets that are a collection of tarred datasets (that are a collection of tar files)

**Review by titu1994 — APPROVED** [review]
Looks great ! We need to update a few of the configs in another pr. Ready to merge

**Comment by titu1994** [comment]
Is this ready to merge ?

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache LiceConnse, Version 2.0 (the "License");
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_text_dataset.py:3]: ?

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 3]
```diff
@@ -94,6 +94,13 @@ def get_concat_char_dataset(
 
     manifest_filepaths = config['manifest_filepath']
     datasets = []
+
+    # needed to support validation Concat Datasets that arrive here as
+    # [[dataset1,dataset2]] otherwise ModelPT would interfere
+    if len(manifest_filepaths) == 1 and not isinstance(manifest_filepaths[0], str):
+        logging.warning(f"flattening {manifest_filepaths}")
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_text_dataset.py:101]: Warning -> info and more descriptive info

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 4]
```diff
@@ -168,6 +178,13 @@ def get_concat_bpe_dataset(
     """
     manifest_filepaths = config['manifest_filepath']
     datasets = []
+
+    # needed to support validation Concat Datasets that arrive here as
+    # [[dataset1,dataset2]] otherwise ModelPT would interfere
+    if len(manifest_filepaths) == 1 and not isinstance(manifest_filepaths[0], str):
+        logging.warning(f"flattening {manifest_filepaths}")
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_text_dataset.py:185]: Same as above

File: nemo/collections/common/data/dataset.py [hunk 5]
```diff
@@ -143,6 +156,8 @@ def temperature_generator(datasets, **kwargs):
         if not temp:
             raise ValueError("Temperature generator expects a 'temperature' keyword argument.")
 
+        seed = kwargs.get('seed')
```
Comment (titu1994) [inline nemo/collections/common/data/dataset.py:159]: How about seed, none as default

**Review by titu1994 — COMMENTED** [review]
Overall it looks a lot better. Minor comments only

File: nemo/collections/common/data/dataset.py (hunk 5 above)
Comment (bmwshop) [reply nemo/collections/common/data/dataset.py]: Done

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 4 above)
Comment (bmwshop) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: done

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 3 above)
Comment (bmwshop) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: done

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 2 above)
Comment (bmwshop) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: fixed

**Review by titu1994 — APPROVED** [review]
Thanks ! LGTM

**Pull request merged**